---
title: "Tutorial - Compare results of Dada2 vs Mothur"
author: "Daniel Vaulot"
date: '`r format(Sys.time(), "%d %m %Y")`'
header-includes:
   - \usepackage{color, fancyvrb}
output:
  rmdformats::readthedown:
    highlight: kate
    number_sections : yes    
  pdf_document: 
    toc: yes
    toc_depth: 2
    number_sections : yes
---

```{r setup, echo=FALSE}
  knitr::opts_chunk$set(fig.width=6, 
                        fig.height=6, 
                        eval=TRUE, 
                        cache=TRUE,
                        echo=TRUE,
                        prompt=FALSE,
                        tidy=TRUE,
                        comment=NA,
                        message=FALSE,
                        warning=FALSE)

```

# Aim

Compare the results of the analysis by mothur vs dada2   

* **Prerequisite** : the R_dada2_tutorial must have been run before

# Directory structure

Relative to the main directory from GitHub

* **../fastq** : fastq files
* **../dada2** : dada2 processed files
* **../R_dada2** : This tutorial


# Setup

## Load the necessary libraries

```{r, results='hide', message=FALSE}
  library("phyloseq") 
  library("Biostrings")
  
  library("ggplot2")
  library("stringr")
  library("dplyr")
  library("tidyr")
  library("readxl")
  library("tibble")
  library("readr")

  library("kableExtra") # necessary for nice table formatting with knitr
```

## Set up directories

```{r}

  fastq_dir <-    "../fastq/"           # fastq directory
  dada2_dir <- "../dada2/"              # dada2 results
```

## PR2 tax levels

```{r}
  PR2_tax_levels <- c("Kingdom", "Supergroup","Division", "Class", 
                      "Order", "Family", "Genus", "Species")
```




# Phyloseq



## Read phyloseq object for dada2

The file has been created in the R_dada2_tutorial.Rmd

```{r}
  ps_dada2<- readRDS(str_c(dada2_dir, "CARBOM_phyloseq.rds"))
```

## Create a phyloseq object for the mothur results

### Sample names

It is assumed that the sample names are at the start of file name and separated by _.

```{r}
# get a list of all fastq files in the ngs directory and separate R1 and R2
  fns <- sort(list.files(fastq_dir, full.names = TRUE)) 
  fns <- fns[str_detect( basename(fns),".fastq")]
  fns_R1 <- fns[str_detect( basename(fns),"R1")]
  fns_R2 <- fns[str_detect( basename(fns),"R2")]

# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
  sample.names <- str_split(basename(fns_R1), pattern = "_", simplify = TRUE) 
  sample.names <- sample.names[,1]

  samdf <- data.frame(sample_name=sample.names)
  rownames(samdf) <- sample.names


```

### Mothur data from Excel file

```{r}
# read the mothur database from Excel file
  mothur_file <- "../Comparison different methods 1.0.xlsx"
  mothur_database <- read_excel(mothur_file, sheet="mothur")
# need to remove empty rows
  mothur_database <- mothur_database %>% filter(!is.na(OTUNumber))

# create the taxonomy matrix  
  mothur_tax <- str_split(mothur_database$OTUConTaxonomy, ";", 9, simplify = TRUE)
# the last column is empty so remove  
  mothur_tax <- mothur_tax[,1:8 ]   
  
# give row names and column names
  colnames(mothur_tax) <- PR2_tax_levels
  rownames(mothur_tax) <- mothur_database$OTUNumber 
  
# replace in the database, the unique taxonomy column by 8 columns
  mothur_database <- cbind(mothur_database, mothur_tax)
  mothur_database <- mothur_database %>% select(-OTUConTaxonomy)
 
# create the otu_table
  mothur_otu <- select(mothur_database, ends_with("p"),-Supergroup)
  rownames(mothur_otu) <- mothur_database$OTUNumber
  mothur_otu <- as.matrix(mothur_otu)

# create the phyloseq for mothur  
  ps_mothur <- phyloseq(otu_table(mothur_otu, taxa_are_rows=TRUE), 
               sample_data(samdf), 
               tax_table(mothur_tax))
```


# Comparison between dada2 and mothur

## Compare at the division level

```{r}
  # remove Lobosa only found in Mothur
  ps_mothur_common <- subset_taxa(ps_mothur, !(Division %in% c("Lobosa")))
  
  plot_bar(ps_mothur_common, fill = "Division") + 
    geom_bar(aes(color=Division, fill=Division), stat="identity", position="stack") +
    ggtitle("Mothur")
  plot_bar(ps_dada2, fill = "Division") + 
    geom_bar(aes(color=Division, fill=Division), stat="identity", position="stack") +
    ggtitle("Dada2")


```
## Compare by aggregation

### Transform the database files into the long version

```{r}
  dada2_database <- read_tsv(str_c(dada2_dir, "CARBOM_dada2.database.tsv"))
  dada2_long <- dada2_database %>% gather(key = "sample", value="n_seq", num_range(120:200,"p"))  %>%
                                   select(-sequence)
  knitr::kable(head(dada2_long), "latex") %>%
         kable_styling(bootstrap_options = "striped", font_size = 7)

  mothur_long <- mothur_database %>% gather(key = "sample", value="n_seq", num_range(120:200,"p"))  %>%
                                   select(-contains("repSeq"), -nseq)
  knitr::kable(head(mothur_long), "latex")  %>%
         kable_styling(bootstrap_options = "striped", font_size = 7)


```

### Aggregate by Division, Class, Genus, Species

```{r}
  dada2_species <- dada2_long %>% group_by(Division, Class, Genus, Species) %>%
                                   summarize (n_seq = sum(n_seq))
  knitr::kable(head(dada2_species))

  mothur_species <- mothur_long %>% group_by(Division, Class, Genus, Species) %>%
                                   summarize (n_seq = sum(n_seq))
  knitr::kable(head(mothur_species))

```

### Merge the two lists to comute the relation between mothur and dada2 estimates

```{r, warning=FALSE}
  both_species <- rbind(dada2_species, mothur_species) %>% 
                        group_by(Division, Class, Genus, Species) %>%
                        summarize (n_methods = n()) %>%
                        left_join(dada2_species) %>%
                        dplyr::rename (dada2 = n_seq) %>%
                        left_join(mothur_species) %>%
                        dplyr::rename (mothur = n_seq) %>%
                        arrange(desc(dada2))
  both_class <- both_species %>%  group_by(Division, Class) %>%
                                  summarise(dada2=sum(dada2, na.rm = TRUE), mothur=sum(mothur, na.rm = TRUE))%>%
                                  arrange(desc(dada2))
```

### Plot at Class level

```{r}
  
  
  ggplot(both_class) + geom_point(aes(mothur,dada2)) +
                         geom_smooth(aes(mothur,dada2), method = "lm", show.legend = TRUE) +
                         ggtitle ("Class level count")
  
  summary(lm(both_class$dada2 ~ both_class$mothur))

```
  
### Plot at Species level


```{r}


  ggplot(both_species) + geom_point(aes(mothur,dada2)) +
                       geom_smooth(aes(mothur,dada2), method = "lm", show.legend = TRUE) +
                       ggtitle ("Species level count")

  summary(lm(both_species$dada2 ~ both_species$mothur))
    
  species_one_method <- both_species %>% filter(n_methods==1) 
  knitr::kable(species_one_method, "latex")%>%
         kable_styling(bootstrap_options = "striped", font_size = 7)

```


# Conclusion on the dada2 pipeline

* The dada2 pipeline yieds 1.7 more reads than mothur
* The number of reads at the species and class levels are correlated
* It is very fast, the longest step is the taxonomy assignement
* It offers the advantage of having everything performed under R. 

